pkg/services/provisioning/provisioning.go

```diff
@@ -13,35 +14,120 @@ import (
-	if err := notifiers.Provision(alertNotificationsPath); err != nil {
-		return fmt.Errorf("Alert notification provisioning error: %v", err)
+type provisioningServiceImpl struct {
+	Cfg                     *setting.Cfg `inject:""`
+	log                     log.Logger
+	dashProvisionerChan     chan dashboards.DashboardProvisioner
+	pollingCtxCancel        context.CancelFunc
+	newDashboardProvisioner dashboards.DashboardProvisionerFactory
+	provisionNotifiers      func(string) error
+	provisionDatasources    func(string) error
+}
+
+func (ps *provisioningServiceImpl) Init() error {
+	err := ps.ProvisionDatasources()
+	if err != nil {
+		return err
+	}
+
+	err = ps.ProvisionNotifications()
+	if err != nil {
+		return err
+	}
+
+	err = ps.ProvisionDashboards()
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func (ps *ProvisioningService) Run(ctx context.Context) error {
+func (ps *provisioningServiceImpl) Run(ctx context.Context) error {
+	for {
+		// We do not have to check for cancellation of the pollingContext in the select statement as there is
+		// nothing to do with that, just wait for new provisioner to start polling again.
+		select {
+		case provisioner := <-ps.dashProvisionerChan:
+			// There is new dashboard provisioner config so we need to start polling for it.
+			// Make sure the old polling process is cancelled even though this should be called before by the code
+			// creating new provisioner to prevent race conditions.
+			ps.cancelPolling()
+
+			pollingContext, cancelFun := context.WithCancel(ctx)
+			ps.pollingCtxCancel = cancelFun
+
+			provisioner.PollChanges(pollingContext)
+		case <-ctx.Done():
+			// Root server context was cancelled so just leave.
+			return ctx.Err()
+		}
+	}
+}
+
+func (ps *provisioningServiceImpl) ProvisionDatasources() error {
+	datasourcePath := path.Join(ps.Cfg.ProvisioningPath, "datasources")
+	err := ps.provisionDatasources(datasourcePath)
+	return errors.Wrap(err, "Datasource provisioning error")
+}
+
+func (ps *provisioningServiceImpl) ProvisionNotifications() error {
+	alertNotificationsPath := path.Join(ps.Cfg.ProvisioningPath, "notifiers")
+	err := ps.provisionNotifiers(alertNotificationsPath)
+	return errors.Wrap(err, "Alert notification provisioning error")
+}
+
+func (ps *provisioningServiceImpl) ProvisionDashboards() error {
 	dashboardPath := path.Join(ps.Cfg.ProvisioningPath, "dashboards")
-	dashProvisioner := dashboards.NewDashboardProvisioner(dashboardPath)
+	dashProvisioner, err := ps.newDashboardProvisioner(dashboardPath)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create provisioner")
+	}
+	// Lets cancel first so we do not get new config temporary overwritten by the polling process.
+	ps.cancelPolling()
 
-	if err := dashProvisioner.Provision(ctx); err != nil {
-		return err
+	if err := dashProvisioner.Provision(); err != nil {
+		return errors.Wrap(err, "Failed to provision dashboards")
 	}
+	ps.dashProvisionerChan <- dashProvisioner
```

Comment:
I've struggling to decide if I like this or not :)
I like the fact that you want to make sure only the latest provisioning config is used. But I can't help to wonder if this would have been cleaner with some locks. I think that would remove the need for `time.Sleep` in the tests as well.

Reply:
Hmm I did not consider that. It is some time I worked with any synchronisation locks so not a first solution that come to my mind. Will take a look at that approach, even if only to make myself familiar with it.

Reply:
I tried the mutex approach, you can see it in the last commit. It seems a bit simpler, on the other hand it still needs waiting for channel + the mutex so not sure it is less complex overall. But it has the benefit that now multiple calls to the reload API won't overlap in nondeterministic way (I think) which could have happened before.